### PR TITLE
ARROW-1572: [C++] Implement "value counts" kernels for tabulating value frequencies

### PR DESCRIPTION
**Pull request by andrioni**
From the tests it seems to work, although I'm still learning how everything works. The implementation is based on how the other hash kernels work.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/hash.cc [hunk 1]
```diff
@@ -871,6 +915,49 @@ Status GetDictionaryEncodeKernel(FunctionContext* ctx,
...
+    COUNT_VALUES_CASE(DoubleType);
+    COUNT_VALUES_CASE(Date32Type);
+    COUNT_VALUES_CASE(Date64Type);
+    COUNT_VALUES_CASE(Time32Type);
+    COUNT_VALUES_CASE(Time64Type);
+    COUNT_VALUES_CASE(TimestampType);
+    COUNT_VALUES_CASE(BinaryType);
+    COUNT_VALUES_CASE(StringType);
+    COUNT_VALUES_CASE(FixedSizeBinaryType);
+    COUNT_VALUES_CASE(Decimal128Type);
```
Comment (wesm) [inline cpp/src/arrow/compute/kernels/hash.cc]: Is there a way to promote some code reuse with the other unary (single-argument) hash kernels?

Comment (andrioni) [reply cpp/src/arrow/compute/kernels/hash.cc]: I think so, although I am not sure how to do it. Maybe moving everything to a macro? I am willing to try if somebody could give me some pointers on what's the best way to do it.

Comment (emkornfield) [reply cpp/src/arrow/compute/kernels/hash.cc]: The pattern described here might be useful: https://mortoray.com/2014/09/10/using-macros-to-simplify-type-visitors-and-enums/

**Review by emkornfield — COMMENTED** [review]
I guess it is redundant information since Null count is already in the array, but it might be good to either:
1.  Clarify in documentation that null elements are not included in counts
2.  Have an option to include them
3.  Always include them.
(1 seems like a reasonable thing for now with a follow-up JIRA)

**Comment by emkornfield** [comment]
Hi @andrioni I'm trying to triage some old bugs and do first pass code reviews.  Thanks for the contribution, I think with a rebase it looks ok to me.  @xhochy it looks like this JIRA is assigned to you, was there any particular reason for that?

**Comment by xhochy** [comment]
@emkornfield I once wanted to implement this but never got to it. :8

**Comment by wesm** [comment]
This PR may need quite a bit of work since @pitrou has revamped the hash kernels since this was written

**Comment by emkornfield** [comment]
@andrioni do you think you would be able to update the pull request for this release?  If not would you mind if it take over?

**Comment by emkornfield** [comment]
@andrioni I'm going to try to pickup were you left off.

**Comment by emkornfield** [comment]
I picked up this up in #3579

**Comment by andrioni** [comment]
Sorry for the delay in answering, and thanks for picking it up!

**Pull request closed without merging**